### PR TITLE
EZ vs EE model, with linear transformation of effects

### PR DESCRIPTION
**Pull request by gaow**
This pull request combines input from @zouyuxin and @gaow implementing EZ vs EE model and tests for difference of the same effect under different conditions (and dealing with EZ vs EE under this context). 

@pcarbo as we've discussed the other day, it is best that you review the pull request, in particular finalizing the interface and improvements RE best practice (I've worked with Yuxin on some best practice issue but you certainly will spot more). To facilitate the process, Yuxin provides a [write-up](https://github.com/stephenslab/mashr/files/1465523/WriteUp.pdf) to summarize changes made. We can discuss on slack. After the merger we'll work on the Rcpp implementation.

### CODE REVIEW

File: R/mash.R [hunk 1]
```diff
@@ -253,6 +334,8 @@ expand_cov = function(Ulist,grid,usepointmass=TRUE){
 #' \code{Shat}, an n by R matrix of standard errors (n units in R
 #' conditions),
 #'
+#' @param alhpa Numeric value of alpha parameter in the model. alpha = 0 for Exchangeable Effects (EE), alpha = 1 for Exchangeable Z-scores (EZ).
```
Comment (stephens999) [inline R/mash.R]: typo here

File: R/posterior.R [hunk 2]
```diff
@@ -73,20 +75,37 @@ posterior_mean_matrix <- function(Bhat, Vinv, U1){
...
+    stop("FIXME: not implemented")
+  }
+  # If A is missing, set A be identity matrix
+  R = n_conditions(data)
+  if(missing(A)){
+    A = diag(R)
+    row.names(A) = colnames(data$Bhat)
+  }
+  if(ncol(A) != R){
+    stop('A is not a propor transformation')
```
Comment (stephens999) [inline R/posterior.R]: typo propor -> proper

**Comment by gaow** [comment]
there should not be a need to keep `data$alpha`, as we do keep both `data$Shat_alpha` and `data$Shat` in data ...

`g$alpha` needs a bit of a discussion: if we do not store `alpha` in `g`, then when in a modular procedure we use `mash_compute_posterior_matrices` with provided `g` yet "vanilla" data input, we'll need to additionally provide an `alpha` that matches the `alpha` when `g` was computed to properly transform the data. It could be confusing to require that match.

**Comment by gaow** [comment]
To summarize out meeting (please add to this list as you see fit):

## Major

- Move data adjustment with `alpha` inside `mash()` and some other function calls (`mash_compute_*`) to `set_mash_data`
   - keep `alpha` value there
   - `data` will then contain `Bhat`, `Shat`, `Shat_alpha`, `alpha`. Though we might rename `Bhat` and `Shat` down the line.
- Move `alpha` out of `fitted_g`, into the `mash` object (outcome of mash analysis). 
- Now that we have 2 places to store `alpha`, we 
  - compare and handle inconsistency of `data$alpha` vs `mash$alpha` when the `mash` object is used to initialize computations
  - when `fitted_g` is used with `data`, we warn that users should be responsible that `data$alpha` is what was used to compute the `fitted_g`
- Merge `posterior_common` and `posterior_lowmem`. Make optimization after `common_cov` check a special case in the more general function call -- because now we have to deal with both `Shat` and `Shat_alpha` having equal elements or not, we have to deal with 4 situations instead of 2; it then seems easier to consolidate to a general version.

**Comment by stephens999** [comment]
Another minor is that I think the two different uses of 'is_common_cov' should be 2 different functions rather than a switch within a single function, as they really have different logic

File: R/set_data.R [hunk 3]
```diff
@@ -34,7 +35,19 @@ set_mash_data = function(Bhat,Shat=NULL,df=NULL,pval=NULL,V=diag(ncol(Bhat))){
     Shat_orig = Shat
```
Comment (stephens999) [inline R/set_data.R]: this line seems superfluous?

**Comment by stephens999** [comment]
this is looking great!

**Comment by stephens999** [comment]
I have pulled these changes into a new branch gaow-master in preparation for merging. I'll close this now

**Pull request closed without merging**